core/capabilities/integration_tests/keystone/keystone_test.go

```diff
@@ -121,6 +123,33 @@ func waitForConsumerReports(t *testing.T, consumer *feeds_consumer.KeystoneFeeds
 	}
 }
 
+// trackErrorsOnForwarder watches the forwarder contract for report processed events and fails the test if the report is not forwarded to the consumer
+func trackErrorsOnForwarder(t *testing.T, forwarder *fwd.KeystoneForwarder, consumerAddress common.Address) {
+	reportsProcessed := make(chan *fwd.KeystoneForwarderReportProcessed, 1000)
+	reportsSub, err := forwarder.WatchReportProcessed(nil, reportsProcessed, nil, nil, nil)
+	require.NoError(t, err)
+
+	ctx := t.Context()
+	go func() {
+		for {
```

Comment:
It is important to block test completion on this goroutine exiting, otherwise it could log after the cancellation signal is sent, and end up causing a data race in the test logs. Something like this should work:
```suggestion
func trackErrorsOnForwarder(t *testing.T, forwarder *fwd.KeystoneForwarder, consumerAddress common.Address) (wait func()) {
	reportsProcessed := make(chan *fwd.KeystoneForwarderReportProcessed, 1000)
	reportsSub, err := forwarder.WatchReportProcessed(nil, reportsProcessed, nil, nil, nil)
	require.NoError(t, err)

	done := make(chan struct{})
	wait := func() { <-done }

	ctx := t.Context()
	go func() {
		defer close(done)
		for {
```
Then the caller just has to do:
```diff
- trackErrorsOnForwarder(t, forwarder, consumer.Address())
+ wait := trackErrorsOnForwarder(t, forwarder, consumer.Address())
+ t.Cleanup(wait)
```

Reply:
Probably cleaner to just return the closure at the end of the func.

Reply:
Good one, I've added it [here](https://github.com/smartcontractkit/chainlink/pull/18841/commits/abfdd57223c333825e2ce709ce0e478454aa4bbb). Since we have `t` in the function, I just called `t.Cleanup()` there, easier for the callers

Reply:
ah, duh :+1: